how-to/workspace-platform-starter/client/src/framework/favorite.ts

```diff
@@ -0,0 +1,210 @@
+import { fireLifecycleEvent } from "./lifecycle";
+import { createLogger } from "./logger-provider";
+import type { EndpointProvider } from "./shapes";
+import type {
+	EndpointFavoriteGetRequest,
+	EndpointFavoriteGetResponse,
+	EndpointFavoriteListRequest,
+	EndpointFavoriteListResponse,
+	EndpointFavoriteRemoveRequest,
+	EndpointFavoriteSetRequest,
+	FavoriteEntry,
+	FavoriteInfo,
+	FavoriteProviderOptions,
+	FavoriteTypeNames
+} from "./shapes/favorite-shapes";
+import type { VersionInfo } from "./shapes/version-shapes";
+import { isEmpty } from "./utils";
+
+const logger = createLogger("Favorites");
+
+let isInitialized: boolean = false;
+
+let favoriteOptions: FavoriteProviderOptions | undefined;
+
+let endpoints: EndpointProvider;
+
+let currentVersionInfo: VersionInfo;
+
+const FAVORITE_ENDPOINT_ID_LIST = "favorite-list";
+const FAVORITE_ENDPOINT_ID_GET = "favorite-get";
+const FAVORITE_ENDPOINT_ID_SET = "favorite-set";
+const FAVORITE_ENDPOINT_ID_REMOVE = "favorite-remove";
+
+/**
+ * Initialize the favorite provider.
+ * @param options Options for the favorite provider.
+ * @param versionInfo The current version information for this platform.
+ * @param endpointProvider The endpoint provider to help retrieve favorites from endpoints.
+ * @returns Nothing.
+ */
+export async function init(
+	options: FavoriteProviderOptions | undefined,
+	versionInfo: VersionInfo,
+	endpointProvider: EndpointProvider
+): Promise<void> {
+	if (isInitialized) {
+		logger.warn("The favorite service is already initialized");
+		return;
+	}
+	currentVersionInfo = versionInfo;
+	endpoints = endpointProvider;
+	if (!isEmpty(options)) {
+		isInitialized = true;
+		favoriteOptions = options;
+	} else {
+		logger.warn("Options were not passed to the favorite provider so it cannot be initialized.");
+	}
+}
+
+/**
+ * Lets you get favorites that have been saved.
+ * @param byType do you wish to filter so it only returns a particular type of favorite
+ * @returns FavoriteEntry[]
+ */
+export async function getSavedFavorites(byType?: FavoriteTypeNames): Promise<FavoriteEntry[]> {
+	if (endpoints.hasEndpoint(FAVORITE_ENDPOINT_ID_LIST)) {
+		const favorites = await endpoints.requestResponse<
+			EndpointFavoriteListRequest,
+			EndpointFavoriteListResponse
+		>(FAVORITE_ENDPOINT_ID_LIST, {
+			favoriteType: byType,
+			platform: fin.me.identity.uuid
+		});
+		if (favorites) {
+			if (!isEmpty(byType)) {
+				logger.info(`Returning saved favorites by type ${byType} from custom storage`);
+			} else {
+				logger.info("Returning saved favorites from custom storage");
+			}
+			return Object.values(favorites).map((fav) => fav.payload);
+		}
+		logger.warn("No response getting saved favorites from custom storage");
+		return [];
+	}
+	logger.warn(`There has been a request for a list of favorites but no endpoint with the 
+    id: ${FAVORITE_ENDPOINT_ID_LIST} is listed in your endpoint configuration.`);
+	return [];
```

Comment:
This should probably return undefined, which is a distinct different value from an empty array, which means there are no favorites

Reply:
yeah, was trying to stay consistent with the way workspaces worked but yeah can return undefined here.

Reply:
done